server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/MergedPointRangeQuery.java

```diff
@@ -0,0 +1,201 @@
+     * that matches points that match both filters.
+     */
+    public static Query merge(PointRangeQuery lhs, PointRangeQuery rhs) {
+        if (lhs.equals(rhs)) {
+            // Lucky case! The queries were the same so their UNION is just the query itself.
+            return lhs;
+        }
+        if (lhs.getField() != rhs.getField() || lhs.getNumDims() != rhs.getNumDims() || lhs.getBytesPerDim() != rhs.getBytesPerDim()) {
+            return null;
+        }
+        return new MergedPointRangeQuery(lhs, rhs);
+    }
+
+    private final String field;
+    private final Query delegateForMultiValuedSegments;
+    private final Query delegateForSingleValuedSegments;
+
+    private MergedPointRangeQuery(PointRangeQuery lhs, PointRangeQuery rhs) {
+        field = lhs.getField();
+        delegateForMultiValuedSegments = new BooleanQuery.Builder().add(lhs, Occur.MUST).add(rhs, Occur.MUST).build();
+        int numDims = lhs.getNumDims();
+        int bytesPerDim = lhs.getBytesPerDim();
+        this.delegateForSingleValuedSegments = pickDelegateForSingleValuedSegments(
+            mergeBound(lhs.getLowerPoint(), rhs.getLowerPoint(), numDims, bytesPerDim, true),
+            mergeBound(lhs.getUpperPoint(), rhs.getUpperPoint(), numDims, bytesPerDim, false),
+            numDims,
+            bytesPerDim
+        );
+    }
+
+    private Query pickDelegateForSingleValuedSegments(byte[] lower, byte[] upper, int numDims, int bytesPerDim) {
+        // If we ended up with disjoint ranges in any dimension then on single valued segments we can't match any docs.
+        for (int dim = 0; dim < numDims; dim++) {
+            int offset = dim * bytesPerDim;
+            if (compareUnsigned(lower, offset, offset + bytesPerDim, upper, offset, offset + bytesPerDim) > 0) {
+                return new MatchNoDocsQuery("disjoint ranges");
+            }
+        }
+        // Otherwise on single valued segments we can only match docs the match the UNION of the two ranges.
+        return new PointRangeQuery(field, lower, upper, numDims) {
+            @Override
+            protected String toString(int dimension, byte[] value) {
+                // Stolen from Lucene's Binary range query. It'd be best to delegate, but the method isn't visible.
+                StringBuilder sb = new StringBuilder();
+                sb.append("(");
+                for (int i = 0; i < value.length; i++) {
+                    if (i > 0) {
+                        sb.append(' ');
+                    }
+                    sb.append(Integer.toHexString(value[i] & 0xFF));
+                }
+                sb.append(')');
+                return sb.toString();
+            }
+        };
+    }
+
+    @Override
+    public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) throws IOException {
+        Weight delegateForMultiValuedSegmentsWeight = delegateForMultiValuedSegments.createWeight(searcher, scoreMode, boost);
```

Comment:
I think we don't need to create this weight here but do it lazy in the same way we are doing for singleValue.

Reply:
:+1: